### ISSUE TITLE
[Bug]: Message "Error running flutter sdk-tools"

### ISSUE THREAD

**Issue by justyn** [akinsho/flutter-tools.nvim#146]
**Describe the bug**

When I open a dart file, it presents the message: "Error running flutter sdk-path"

**To Reproduce**

Open a dart file eg "widget_test.dart"

**Expected behavior**

No warning presented.

**System information**

- OS: Ubuntu 20.04
- Flutter version: 2.10.5
- Is flutter in $PATH: Yes
- neovim version: v0.6.0 (v0.8-dev also tried)

**Additional context**

flutter and neovim are installed as snaps.

`flutter sdk-path` runs in the terminal and presents the output:

`/home/MYUSERNAME/snap/flutter/common/flutter`

This is my first time setting up neovim, apologies if it is an obvious oversight.

init.lua follows:

```
require('packer').startup(function(use)
  use 'wbthomason/packer.nvim'
  use {'akinsho/flutter-tools.nvim', requires = 'nvim-lua/plenary.nvim'}
end) 

require("flutter-tools").setup{}
```

**Comment by akinsho** [comment]
@justyn I think the issue might be that you are running your setup function for this plugin at the wrong time, which is an easy mistake to make with packer. I'm a mobile currently so I can't do a code block easily but essentially rather than calling set up immediately after calling use, you should add a config key Which should be a function and inside of it is where you call setup of this plugin. Have a look at Packer's documentation for how config works

**Comment by justyn** [comment]
Hi @akinsho 

Thanks, and yes I completely believe this is user error since I haven't used nvim before.

Currently packer and flutter-tools are the only plugins I'm installing.

I've adjusted the flutter tools section to be this:

```
    use {                                                                                                                                                     
        'akinsho/flutter-tools.nvim',                                                                                                                         
        requires = 'nvim-lua/plenary.nvim',                                                                                                                   
        config = require('flutter-tools').setup{}
    } 
```

However, I'm still seeing the message, sorry.

**Comment by akinsho** [comment]
@justyn you need to adjust your config a bit more to
```lua
    use {                                                                                                                                                     
        'akinsho/flutter-tools.nvim',                                                                                                                         
        requires = 'nvim-lua/plenary.nvim',                                                                                                                   
        config = function()
			require('flutter-tools').setup{}
		end
    } 
```
The key here is that you are trying to delay calling `require('flutter-tools').setup()` to later in nvim's initialization, whereas without the function wrapper you were calling it immediately

**Comment by justyn** [comment]
@akinsho so sorry but I've just tried your adjustment, and run :Packer{Compile,Install,Sync} a few times for good measure, and restarted nvim, and I'm still seeing the same error.

I wonder if it's actually something to do with my system somehow?

**Comment by justyn** [comment]
I deleted my `~/.config/neovim` directory and started again.

It contains only `init.lua`, pasted below in its entirety:

```
require('packer').startup(function(use)
  use 'wbthomason/packer.nvim'
  use {                                                                                                                                                     
        'akinsho/flutter-tools.nvim',                                                                                                                         
        requires = 'nvim-lua/plenary.nvim',                                                                                                                   
        config = function()
		require('flutter-tools').setup{}
		end
    }
end) 
```

Then ran `:PackerSync` which worked with no issues.

But same error. So I feel like something else is going on.

Do you have any pointers on how I might debug?

`flutter sdk-path` returns correct path in terminal (same terminal I'm running nvim from`).

Thanks

**Comment by akinsho** [comment]
@justyn I'm not really sure, I don't run Linux any more, so I can't check for you, but as you say it might be something about how things are setup on your system. The best guess I have is adding a print statement where the fetch is being done to see why it's failing, can you also confirm for me that `plenary` is correctly installed and available e.g.`lua =require('plenary')` should return a table

**Comment by akinsho** [comment]
Re. adding the log you can add the argument job to the function call `function(job)` 
then print `vim.prettyprint(job:result())` and paste the result here.

https://github.com/akinsho/flutter-tools.nvim/blob/e9471a942900d18f4cf2376400a36f8592321d52/lua/flutter-tools/executable.lua#L82

**Comment by justyn** [comment]
Thank you @akinsho I will investigate along those lines as soon as I am able.

**Comment by justyn** [comment]
Hi @akinsho I'm new to lua and couldn't work out `vim.prettyprint`, but just adding `job` argument to the schedule_wrap function, then `print(job:result())` led to this on the display:

```
table: 0x7f737d548bf0
```

I tried converting the table to a string with this code snippet from [SO](https://stackoverflow.com/a/27028488/6835649):

```
function dump(o)
   if type(o) == 'table' then
      local s = '{ '
      for k,v in pairs(o) do
         if type(k) ~= 'number' then k = '"'..k..'"' end
         s = s .. '['..k..'] = ' .. dump(v) .. ','
      end
      return s .. '} '
   else
      return tostring(o)
   end
end
```

then I added this in `job:after_failure`:

```
      io.output("templog")
      io.write(dump(job:result()))
      io.output(io.stdout)
```

But this is all that gets put in the file:

```
{ }
```

Entire function then looks like this:

```
  job:after_failure(vim.schedule_wrap(function(job)
    ui.notify(
      { string.format("Error running %s", lookup_cmd) },
      { timeout = 5000, level = ui.ERROR }
    )
      io.output("biscuit")
      io.write(dump(job:result()))
      io.output(io.stdout)
  end))
```

Not entirely sure if I'm dumping the output incorrectly or if it is just unhelpful.

**Comment by justyn** [comment]
`lua =require('plenary')` results in this:

```
{                                                                                                                                                                   
  <metatable> = {
    __index = <function 1>
  }
}
```

**Comment by justyn** [comment]
Aha, I was looking at some other tool called prettyprint and could not get it to play ball. `vim.pretty_print` works.

It prints: `{}` so that seems to match what I had seen before.

I can't find any other flutter installations on my system, I've only installed it once as far as I know.

The command that is failing in nvim is just `flutter sdk-path` which runs okay in the terminal. Given that this just outputs a string with a path, it's hard to see what could go wrong.

I wonder if plenary.job requires something else to be passed to it on Linux?

**Comment by justyn** [comment]
I've found that an emacs dart plugin also has trouble on Linux systems using snap (the recommended way to install Flutter on Linux): https://github.com/emacs-lsp/lsp-dart/issues/107

Although the fix there seems to be to use `flutter sdk-path` which is the command that is failing here anyway, so perhaps this isn't helpful.

**Comment by akinsho** [comment]
@justyn so I developed this plugin whilst running linux using Ubuntu, so I actually added the sdk-path initially for myself since that was the easiest way to find the binary, so I'm sure that that works since it did for my right up till I switched to a mac recently. I haven't changed that code in a very long time since it tends to cause loads of issues if I change it since someone's machine somewhere will break.

Plenary job shouldn't have any issue resolving that command, it could be that something is maybe erroring

**Comment by akinsho** [comment]
@justyn can you try something similar but with
```lua
  job:after_failure(vim.schedule_wrap(function(...)
    vim.pretty_print(...) -- <- This will print all the arguments passed to the function
    ui.notify(
      { string.format("Error running %s", lookup_cmd) },
      { timeout = 5000, level = ui.ERROR }
    )
  end))
```

**Comment by justyn** [comment]
@akinsho 

I got this output with the ...

```
{
	["is_shutdown"] = true,
	["command"] = "flutter",
	["_stdout_results"] = 
	{
	},
	["stdout"] = uv_pipe_t: 0x558887bdd2a0,
	["stderr"] = uv_pipe_t: 0x558887bdfc40,
	["pid"] = 177360,
	["code"] = 1,
	["interactive"] = true,
	["handle"] = uv_process_t: 0x558887bdc420,
	["enable_handlers"] = true,
	["enable_recording"] = true,
	["user_data"] = 
	{
	},
	["signal"] = 0,
	["_stderr_results"] = 
	{
	},
	["stdin"] = uv_pipe_t: 0x558887bdfe10,
	["_additional_on_exit_callbacks"] = 
	{
		[1] = function: 0x7f26aef57c58,
		[2] = function: 0x7f26aef57e08,
	},
	["args"] = 
	{
		[1] = "sdk-path",
	},
```

Thanks for the help

**Comment by akinsho** [comment]
Hmm, all that really clarifies is that the command is definitely failing when run from inside neovim, I wonder if you'd get the same error when run inside neovim's terminal 🤔(not that that really matters since it's not exactly equivalent). I think there's a plenary method to view the errors but it's not a super well documented library unfortuntely

**Comment by akinsho** [comment]
Ah found another example from somewhere else in the code base, can you print `j:stderr_result()` i.e.
```
job:after_failure(vim.schedule_wrap(function(j)
	vim.pretty_print(j:stderr_result())
-- ...
end))
```

**Comment by justyn** [comment]
@akinsho seems to be empty, just shows `{}`

**Comment by justyn** [comment]
I did just try running `flutter sdk-path` in neovims `:terminal` and it worked, returned the correct path. 🤷

**Comment by justyn** [comment]
Just as a sanity check I changed `lookup_cmd` to `which flutter` and that executed correctly (and returned `/snap/bin/flutter` as expected.

**Comment by akinsho** [comment]
This is really quite/hard for me to test on mac, so I can't really dig any deeper, and jumping around in VMs is a step beyond what I'm keen on doing, since my nvim plugins are already quite a large time sink I never planned on.

I guess if `which flutter` works for you, you can use that. I'm guessing that if this was a general problem, I'd have seen more reports of this. I've had quite a few issues like this about finding the flutter path across a bunch of OSes and distros etc. which is why there are so many mechanisms to find or specify the path.

I guess one other thing to do is to try checking the `j:stdout_result()` and maybe also logging inside the `after_success` handler just to see if anything is going on there.

**Comment by justyn** [comment]
Well unfortunately `which flutter` isn't returning a result that can be used for the SDK, it leads to other errors when the plugin can't find what its looking for. I meant I just ran it to make sure commands were running correctly at that point in the code.

I hear you about the time sink, and thank you for the time already spent on this.

One last thing I would like to try is to bypass the `flutter sdk-path` lookup and provide the result of that command (which is `/home/MYUSER/snap/flutter/common/flutter` directly.

I tried to pass the option `flutter_path` in `init.lua`, which docs say would take priority over the lookup, but I still get the message "Error running flutter sdk-tools".

Could you recommend the best approach to achieve this? Thanks

**Comment by akinsho** [comment]
@justyn I only now just noticed amongst all of this that your `init.lua` is configured incorrectly. You cannot call setup immediately calling `packer.startup` it doesn't work like vim plug so by the time you call setup you probably havent loaded the plugin properly yet.

```lua
require('packer').startup(function(use)
  use 'wbthomason/packer.nvim'
  use {'akinsho/flutter-tools.nvim', requires = 'nvim-lua/plenary.nvim'}
end) 

require("flutter-tools").setup{}
```
You should put your setup function inside of a config key in the use block e.g.
```lua
use {"thisplugin", config = function()
 require('flutter-tools').setup()
end}
```

Can you please run things this way and see if it helps

**Comment by justyn** [comment]
Hi @akinsho sorry maybe I'm getting confused but I think I already have this, my `init.lua` is pasted here: https://github.com/akinsho/flutter-tools.nvim/issues/146#issuecomment-1116039682

Is this still wrong?

**Comment by justyn** [comment]
Well after deleting packer and packer_compiled, I tried again with the latest flutter-tools but this time if I try to open a dart file it says `Error detected while processing FileType Autocommands for "dart" [snip] client_id: expected number, got nil`.

I'm taking this as a sign from the universe to give up for now.

Thanks for spending the time to try to debug my issue. Perhaps I'll revisit it at a later date.

**Comment by justyn** [comment]
It's working! 🎉

The issue I described here remains - "Error running flutter sdk-tools".

If I run this command in a terminal, it gives me the path the the SDK.
By appending `/bin/flutter` to this and passing it to flutter-tools with: `flutter-tools = "PATH` it works.

On my first attempt without appending `/bin/flutter`, I got an error, because flutter-tools was assuming that the directory 2 levels above the path I set was the SDK.

So for future reference by anybody else, this was the minimal init.lua file I used to get it running:

```
require('packer').startup(function(use)
  use 'wbthomason/packer.nvim'
  use {
        'akinsho/flutter-tools.nvim',
        requires = 'nvim-lua/plenary.nvim',
        config = function()
		require("flutter-tools").setup {
			flutter_path = "/home/MYUSER/snap/flutter/common/flutter/bin/flutter",
		}
		end
    }
end) 
```

Thanks again @akinsho for all the help (and of course for developing the plugin).
